zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/ZeppelinHubRepo.java

```diff
@@ -0,0 +1,200 @@
+  private static final Gson GSON = new Gson();
+  private static final Note EMPTY_NOTE = new Note();
+  private final Client websocketClient;
+
+  private String token;
+  private ZeppelinhubRestApiHandler zeppelinhubHandler;
+
+  public ZeppelinHubRepo(ZeppelinConfiguration conf) {
+    String zeppelinHubUrl = getZeppelinHubUrl(conf);
+    LOG.info("Initializing ZeppelinHub integration module");
+    token = conf.getString("ZEPPELINHUB_API_TOKEN", ZEPPELIN_CONF_PROP_NAME_TOKEN, "");
+    zeppelinhubHandler = ZeppelinhubRestApiHandler.newInstance(zeppelinHubUrl, token);
+
+    websocketClient = Client.initialize(getZeppelinWebsocketUri(conf),
+        getZeppelinhubWebsocketUri(conf), token, conf);
+    websocketClient.start();
+  }
+
+  private String getZeppelinHubWsUri(URI api) throws IOException {
+    URI apiRoot = api;
+    String scheme = apiRoot.getScheme();
+    int port = apiRoot.getPort();
+    if (port <= 0) {
+      port = (scheme != null && scheme.equals("https")) ? 443 : 80;
+    }
+
+    if (scheme == null) {
+      LOG.info("{} is not a valid zeppelinhub server address. proceed with default address {}",
+          apiRoot, DEFAULT_SERVER);
+      try {
+        apiRoot = new URI(DEFAULT_SERVER);
+      } catch (URISyntaxException e) {
+        LOG.error("Invalid default zeppelinhub url {}", DEFAULT_SERVER, e);
+        throw new IOException(e);
+      }
+      scheme = apiRoot.getScheme();
+      port = apiRoot.getPort();
+      if (port <= 0) {
+        port = (scheme != null && scheme.equals("https")) ? 443 : 80;
+      }
+    }
+    String ws = scheme.equals("https") ? "wss://" : "ws://";
+    return ws + apiRoot.getHost() + ":" + port + "/async";
+  }
+
+  String getZeppelinhubWebsocketUri(ZeppelinConfiguration conf) {
+    String zeppelinHubUri = StringUtils.EMPTY;
+    try {
+      zeppelinHubUri = getZeppelinHubWsUri(new URI(conf.getString("ZEPPELINHUB_API_ADDRESS",
+          ZEPPELIN_CONF_PROP_NAME_SERVER, DEFAULT_SERVER)));
+    } catch (URISyntaxException | IOException e) {
+      LOG.error("Cannot get zeppelinhub URI", e);
+    }
+    return zeppelinHubUri;
+  }
+
+  private String getZeppelinWebsocketUri(ZeppelinConfiguration conf) {
+    int port = conf.getServerPort();
+    if (port <= 0) {
+      port = 80;
+    }
+    String ws = conf.useSsl() ? "wss" : "ws";
+    return ws + "://localhost:" + port + "/ws";
+  }
+
+  public void setZeppelinhubRestApiHandler(ZeppelinhubRestApiHandler zeppelinhub) {
+    zeppelinhubHandler = zeppelinhub;
+  }
+
+  String getZeppelinHubUrl(ZeppelinConfiguration conf) {
+    if (conf == null) {
+      LOG.error("Invalid configuration, cannot be null. Using default address {}", DEFAULT_SERVER);
+      return DEFAULT_SERVER;
+    }
+    URI apiRoot;
+    String zeppelinhubUrl;
+    try {
+      String url = conf.getString("ZEPPELINHUB_API_ADDRESS",
+                                  ZEPPELIN_CONF_PROP_NAME_SERVER,
+                                  DEFAULT_SERVER);
```

Comment:
Probably a nitpick, but [horizontal alignment](https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment) is something that's not always good